Adding an Interface for DI and CancellationToken for async methods

kdcllc opened the pull request:
- Adding an interface for DI creation of the client not thru the new keyword
- Adding Cancellation Tokens for all of the Async methods to be truly async

kdcllc commented:
@ScottRudiger any chances for this PR to move forward?

codymullins commented on src/Taxjar/Taxjar.csproj:
```diff
@@ -31,4 +31,24 @@
 
+  <ItemGroup Condition="'$(TargetFramework)' == 'netstandard2.0'">
+    <PackageReference Include="Microsoft.Extensions.DependencyInjection.Abstractions" Version="3.1.10"/>
+    <PackageReference Include="Bet.Extensions.Options" Version="3.1.10"/>
```
It would be preferred to strictly use the `Microsoft.Extensions.Options` package here instead.

kdcllc replied:
I can defiantly remove this dependency but fyi this code registers options that will be updated if the underlining configuration provider has changed. This is very helpful in situation where you don't want to restart the application for new options values to be propagated to the app.

codymullins commented:
Thanks for the PR, @kdcllc! We will check it out.

codymullins commented on src/Taxjar/DependencyInjection/TaxjarOptions.cs:
```diff
@@ -0,0 +1,25 @@
+        /// <summary>
+        /// Timeout for the HttpClient.
+        /// </summary>
+        public TimeSpan Timeout { get; set; }
```
Prefer to use `int` directly here, unless you have a compelling case for usage of `TimeSpan`.

kdcllc replied:
What I have experienced from .net side of things it is easier to add a timestamp rather than figure out if it is seconds, min, hour... etc.

codymullins commented on src/Taxjar/DependencyInjection/TaxjarOptions.cs:
```diff
@@ -0,0 +1,25 @@
+        /// <summary>
+        /// Is Sandbox?
+        /// </summary>
+        public bool IsSandBox { get; set; }
```
In a future PR we are intending to change this to accept the URL directly. Since we are adding this, it would probably be better to go ahead and make this `ApiUrl` instead. Then the configuration for projects would be setting the URL instead of `IsSandbox`.

kdcllc replied:
What I have experienced from .net side of things it is easier to add a timestamp rather than figure out if it is seconds, min, hour... etc.

kdcllc replied:
From our testing of the APIs we discovered that Sandbox doesn't support:
- Address Validation
- Customer creation

Thus we ended up creating a factory with Named Options. I am okay with creating `ApiUrl` property that possibly will correspond to Named options.

codymullins commented on src/Taxjar/DependencyInjection/TaxjarServiceCollectionExtensions.cs:
```diff
@@ -0,0 +1,37 @@
+                                                   Action<TaxjarOptions> configure = default)
+        {
+
+            services.AddChangeTokenOptions(nameof(TaxjarOptions), string.Empty, configure);
```
Since this is dependent on the `BET.Extensions.Options` library, this line will need to be removed.

kdcllc replied:
Yes, I can replace with AddOptions<> to register this.

kdcllc commented:
@codymullins I addressed all of the issues that you brought up. Let me know if this PR will do the job. I also at some point recommend enabling specifying ApiVersion via Options.

kdcllc commented:
Any progress made on this?

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/taxjar/taxjar.net?pullRequest=48) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/taxjar/taxjar.net?pullRequest=48) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/taxjar/taxjar.net?pullRequest=48) it.</sub>
